encfs/GCrypt_Cipher.cpp

```diff
@@ -0,0 +1,917 @@
+      cout << "Error in gcry_mac_reset";
+
+    if ((error = gcry_mac_write (key->gcryptMAC, (void *) ivec, _ivLength)) != 0)
+      RLOG(WARNING) << "Error in gcry_mac_write";
+
+    if ((error = gcry_mac_write (key->gcryptMAC, md, 8)) != 0)
+      RLOG(WARNING) << "Error in gcry_mac_write";
+
+    if ((error = gcry_mac_read (key->gcryptMAC, md, &mdLen)) != 0)
+      RLOG(WARNING) << "Error in gcry_mac_write";
+
+
+    rAssert(mdLen >= _ivLength);
+
+    memcpy(ivec, md, _ivLength);
+  } else {
+    setIVec_old(ivec, seed, key);
+  }
+}
+
+/** For backward compatibility.
+    A watermark attack was discovered against this IV setup.  If an attacker
+    could get a victim to store a carefully crafted file, they could later
+    determine if the victim had the file in encrypted storage (without
+    decrypting the file).
+  */
+void GCrypt_Cipher::setIVec_old(unsigned char *ivec, unsigned int seed,
+                             const std::shared_ptr<GCryptKey> &key) const {
+  /* These multiplication constants chosen as they represent (non optimal)
+     Golumb rulers, the idea being to spread around the information in the
+     seed.
+
+     0x060a4011 : ruler length 26, 7 marks, 21 measurable lengths
+     0x0221040d : ruler length 25, 7 marks, 21 measurable lengths
+  */
+  unsigned int var1 = 0x060a4011 * seed;
+  unsigned int var2 = 0x0221040d * (seed ^ 0xD3FEA11C);
+
+  memcpy(ivec, IVData(key), _ivLength);
+
+  ivec[0] ^= (var1 >> 24) & 0xff;
+  ivec[1] ^= (var2 >> 16) & 0xff;
+  ivec[2] ^= (var1 >> 8) & 0xff;
+  ivec[3] ^= (var2)&0xff;
+  ivec[4] ^= (var2 >> 24) & 0xff;
+  ivec[5] ^= (var1 >> 16) & 0xff;
+  ivec[6] ^= (var2 >> 8) & 0xff;
+  ivec[7] ^= (var1)&0xff;
+
+  if (_ivLength > 8) {
+    ivec[8 + 0] ^= (var1)&0xff;
+    ivec[8 + 1] ^= (var2 >> 8) & 0xff;
+    ivec[8 + 2] ^= (var1 >> 16) & 0xff;
+    ivec[8 + 3] ^= (var2 >> 24) & 0xff;
+    ivec[8 + 4] ^= (var1 >> 24) & 0xff;
+    ivec[8 + 5] ^= (var2 >> 16) & 0xff;
+    ivec[8 + 6] ^= (var1 >> 8) & 0xff;
+    ivec[8 + 7] ^= (var2)&0xff;
+  }
+}
+
+static void flipBytes(unsigned char *buf, int size) {
+  unsigned char revBuf[64];
+
+  int bytesLeft = size;
+  while (bytesLeft) {
+    int toFlip = MIN(sizeof(revBuf), bytesLeft);
+
+    for (int i = 0; i < toFlip; ++i) revBuf[i] = buf[toFlip - (i + 1)];
+
+    memcpy(buf, revBuf, toFlip);
+    bytesLeft -= toFlip;
+    buf += toFlip;
+  }
+  memset(revBuf, 0, sizeof(revBuf));
+}
+
+static void shuffleBytes(unsigned char *buf, int size) {
+  for (int i = 0; i < size - 1; ++i) buf[i + 1] ^= buf[i];
+}
+
+static void unshuffleBytes(unsigned char *buf, int size) {
+  for (int i = size - 1; i; --i) buf[i] ^= buf[i - 1];
+}
+
+
+
+/** Partial blocks are encoded with a stream cipher.  We make multiple passes on
+ the data to ensure that the ends of the data depend on each other.
+*/
+bool GCrypt_Cipher::streamEncode(unsigned char *buf, int size, uint64_t iv64,
+                              const CipherKey &ckey) const {
```

Comment:
Done, I implemented the modifications you did in #295 to the SSL_Cipher.cpp file to my GCrypt_Cipher.cpp